### PR TITLE
[Cognitive Services - Azure AI Content Safety] - 2023-10-15-preview Public Preview

### PR DESCRIPTION
**Pull request by suxi-ms**
# Data Plane API - Pull Request

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [ ] a private preview
- [x] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

### CODE REVIEW

**Comment by suxi-ms** [comment]
issue link in private repo: https://github.com/Azure/azure-rest-api-specs-pr/issues/13971

**Review by mikekistler — APPROVED** [review]
Looks good for public preview. 👍

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:30]: Since this is public documentation,

```suggestion
  @doc("Output severities in four levels, the value could be 0, 2, 4, or 6.")
```

Same below.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:236]:
```suggestion
  @doc("Severity less than or equal to the threshold will be treated as accepted for the category. The range is the severity value. Current the value could be: 0, 2, 4, or 6.")
```

Please these same changes to all similar occurances.

**Review by heaths — CHANGES_REQUESTED** [review]
A few nits on doc grammar; otherwise, LGTM.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp:30]: Thanks heaths for the comments. This model is not introduced in change of this PR, but I will fix it .
cc : @mengaims  for awareness.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp:236]: Will do.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp]: Updated, please help review again.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp]: updated

**Review by mikekistler — CHANGES_REQUESTED** [review]
This looks good but let's hold off merging until #25498 merges.

**Comment by suxi-ms** [comment]
> This looks good but let's hold off merging until #25498 merges.

Sure

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp]: Given `JudgementKind` has only 2 values and 1 requires that `severityThreshold` have a value (which itself is optional),

1. Why is `judgementKind` required even though `severityThreshold` is optional?
2. Given `AcceptAll` implies `severityThreshold` is absent (undefined) and `AcceptLessThan` requires a value, why not just use whether `severityThreshold` has a value and then treat it as `AcceptLessThan`?

This design requires more input - and has to be right, which isn't properly expressed here - than just using the existence and, if set, value of `severityThreshold`.

**Review by heaths — CHANGES_REQUESTED** [review]

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp]: Thanks for the feedback. If we use a `null` as for the threshold, user may get the information that no threshold is provided, so the API won't do judgement.

Comment (heaths) [reply specification/cognitiveservices/ContentSafety/models.tsp]: If a threshold is required, then why did you author it as optional? It's odd - and probably not the best UX - to *require* users to author a `judgementKind` when they don't have to author a `severityThreshold`. Instead, maybe you should default `judgementKind` as it corresponds to an absent (different than a `null` value) `severityThreshold`.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:316]: It's in source control, so please just remove this instead.

File: specification/cognitiveservices/data-plane/ContentSafety/preview/2023-10-15-preview/contentsafety.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/ContentSafety/preview/2023-10-15-preview/contentsafety.json:1161]: `-1` seems like such a magic number. Why not just leave the field as optional and, if not specified, treat it like you would a value of -1? That's more RESTful.

**Review by heaths — CHANGES_REQUESTED** [review]

File: specification/cognitiveservices/data-plane/ContentSafety/preview/2023-10-15-preview/contentsafety.json
Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/ContentSafety/preview/2023-10-15-preview/contentsafety.json:1161]: Have removed the -1, and pushed the new changes just as we discussed offline

Comment (heaths) [reply specification/cognitiveservices/data-plane/ContentSafety/preview/2023-10-15-preview/contentsafety.json:1161]: I don't see that reflected in the change you just pushed: https://github.com/Azure/azure-rest-api-specs/pull/25570/commits/314db099dbb3c1a0e053aeec69a378331a78b7c5

Comment (suxi-ms) [reply specification/cognitiveservices/data-plane/ContentSafety/preview/2023-10-15-preview/contentsafety.json:1161]: Hi heaths,I double checked that avaiable judgementKind doesn't exist in model.tsp or 2023-10-15-preview/coontentsafety.json

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:29]: "Kind" is preferred over "Type" since "Type" can mean many different things in SDK languages.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:285]: Similarly, this should be `outputKind`. Same for any other similar use throughout.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:271]: You say, "If the threshold is undefined...", but this property - unlike line 261 above - is required. You're missing a `?` to make this property optional.

**Review by heaths — CHANGES_REQUESTED** [review]
A couple of things to consider, but you need to make the other `acceptSevrityThreshold` optional per conversation.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp:271]: updated

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp]: Just as offline discussed, we have got an alignment that severityThreshold could be null to represent accept all

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp]: Code has been removed

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:40]: Nit: terminal commas are preferred, actually. Makes subsequent diffs smaller and easier to review.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:33]: Nit: more easily read:
```suggestion
  @doc("Output severities in eight levels, the value could be 0, 1, 2, 3, 4, 5, 6, 7.")
```

**Review by heaths — CHANGES_REQUESTED** [review]
Better, but still a couple open feedback items. Can you respond by either fixing them and resolving the comment, or reply why not and resolve so we know when the PR is done?

**Comment by suxi-ms** [comment]
> A couple of things to consider, but you need to make the other `acceptSevrityThreshold` optional per conversation.

I think we have made it optional

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp]: updated

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp]: updated

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (suxi-ms) [reply specification/cognitiveservices/ContentSafety/models.tsp:288]: This concept is to align with the original output type in 10-01 version, which has been GA just yesterday.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (heaths) [inline specification/cognitiveservices/ContentSafety/models.tsp:332]: Avoid acronyms. Not everyone will know what "Nl" means. Even I am only assuming it means "Natural Language". Be verbose. Our SDK architects won't approve any uncommon abbreviations like this.

**Comment by heaths** [comment]
There's been a lot of changes on this, but not all resolved comments seem to be updated in the swagger. Should we expect them to? What is the status on this PR currently? Let @mikekistler and I know so we can (re)do a proper review.

File: specification/cognitiveservices/ContentSafety/models.tsp
Comment (mengaims) [reply specification/cognitiveservices/ContentSafety/models.tsp]: We've updated to ProtectedMaterialAnalysisResult

**Comment by mengaims** [comment]
> There's been a lot of changes on this, but not all resolved comments seem to be updated in the swagger. Should we expect them to? What is the status on this PR currently? Let @mikekistler and I know so we can (re)do a proper review.

Hi @heaths, sorry for the confusion, this PR has been changed a lot since it's open, because there are much releasing plan adjustment from our side. We will avoid this by abandon old PR and creating a new version number and PR if there is big feature release scope change.
For the unresolved comments, I closed the old ones because the Project feature has been abandoned in this version, and for comments since our last [review](https://github.com/Azure/azure-rest-api-specs/issues/25964)(10/26), I've replied and resolved.

We may still hold this PR for one or two weeks to wait our SDK GA code complete, I will ask for your approval when it's ready to merge. Thanks!

**Review by mikekistler — CHANGES_REQUESTED** [review]
Please add
```
omit-unreachable-types: true
```
to the options for typespec-autorest in your tspconfig.yaml.

This will let you avoid the "AddedEnumValue" Swagger breaking change going forward.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍

File: specification/cognitiveservices/ContentSafety/routes.tsp
Comment (weidongxu-microsoft) [inline specification/cognitiveservices/ContentSafety/routes.tsp:40]: for SDK, should you follow GAed APIs and project this `body` to `options` for some languages? Same below.

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
Another SDK comment is that the new operation should be added to client.tsp https://github.com/Azure/azure-rest-api-specs/blob/13a27780750b3130bf7aab374b523093c4560a6a/specification/cognitiveservices/ContentSafety/client.tsp
Else it won't appear at all (e.g. the current automated PR https://github.com/azure-sdk/azure-sdk-for-java/pull/6727/files).

File: specification/cognitiveservices/ContentSafety/routes.tsp
Comment (mengaims) [reply specification/cognitiveservices/ContentSafety/routes.tsp:43]: Added

**Review by heaths — APPROVED** [review]

**Pull request merged**
